### PR TITLE
Examples

### LINKED ISSUE #2700: Examples
Implementation issue for: https://github.com/Azure/typespec-azure/issues/3540, #514.  Design details are [here](https://gist.github.com/markcowl/a5fb8c7048d2ba80fc8f63cd2b97cb8b)

- [ ] Update versioning decorators to accommodate 'never' for the version argument (meaning do not execute the decorator)
- [ ] Add templates and decorators to OpenAPI library
  - [ ] Add JsonExample templates, and, in private namespace,  `@jsonExampleName`, `@jsonExampleLiteral` , `@jsonExampleParameters`, `@jsonExampleReturnValue` decorators and associated accessors:

  ```tsp
    @jsonExampleName(TName)
    @added(TVersionAdded)
    @removed(TVersionRemoved)
    model JsonExampleOptions<
      TName extends string = null,
      TVersionAdded extends EnumMember = never,
      TVersionRemoved extends EnumMember = never> {}

    @jsonExampleLiteral(TExample)
    model JsonExample<
      TExample extends unknown,
      TName extends string = null,
      TSummary extends string = null,
      TVersionAdded extends EnumMember = never,
      TVersionRemoved extends EnumMember = never
    > is ExampleOptions<TName, TVersionAdded, TVersionRemoved> {}

    // This is specifically for matching return values in OpenAPI, with a
    // model keyed by status code
    @jsonExampleParameters(TParams)
    @jsonExampleReturnValue(TReturnValue)
    model JsonOperationExample<
      TParams extends {},
      TReturnValue extends Model | Model[],
      TName extends string = null,
      TSummary extends string = null,
      TVersionAdded extends EnumMember = never,
      TVersionRemoved extends EnumMember = never
    > is ExampleOptions<TName, TVersionAdded, TVersionRemoved> {}
  ```
  - [ ] Add `@jsonExample` decorator and associated accessors
  
  ```tsp
  extern dec jsonExample(target: Model | ModelProperty , example: unknown)
  extern dec jsonOperationExample(target: Operation , example: unknown)
  ```
  - [ ] Add JsonExternalExample templates and associated decorator (`@jsonExamplePath` and accessor in private namespace) 
  
  ```tsp
  @jsonExamplePath(TExternalValue)
  model JsonExternalExample<
    TExternalValue extends string,
    TName extends string,
    TSummary extends string = null,
    TVersionAdded extends EnumMember = never,
    TVersionRemoved extends EnumMember = never
  > is ExampleOptions<TName, TVersionAdded, TVersionRemoved> {}
  ```
  - [ ] Add `@jsonExternalExample` decorator and associated accessors

  ```tsp
  extern dec jsonExternalExample(target: Model | ModelProperty | Operation, externalExample: unknown)
  ```
- [ ] Implement OpenAPI3 example support based on decorators
  - [ ] Model and ModelProperty examples: support inlined value examples and named examples:
  
  ### TypeSpec
  ```tsp
  @jsonExample(
    JsonExample<
      {type: "bar", createdAt: "2023-03-01T19:00:00Z"}, 
      TName = "bar-example", 
      TSummary ="An example of a bar type foo">)
  model Foo {
    type: "bar" | "baz"
    @jsonExample("2023-03-01T19:00:00Z")
    createdAt: utcDateTime;
  }
  ```
  ### Corresponding OpenAPI3
  ```yaml
    Foo:
     type: object
     examples:
       bar-example:
         summary: "An example of a bar type foo"
         value:
           type: "bar"
           createdAt: "2023-03-01T19:00:00Z"
     properties:
       createdAt: 
         type: string
         format: date-time
         examples:
           - "2023-03-01T19:00:00Z"
  ```
  - [ ] Example file support in OpenAPI3
  
  ### TypeSpec
  ```tsp
  @jsonExternalExample(
    JsonExternalExample<
      "examples/simple-foo-example.json", 
      TName = "simple-foo-example", 
      TSummary ="A simple example of type foo">)
  model Foo {
    type: "bar" | "baz"
    createdAt: utcDateTime;
  }
  ```

  ### Corresponding OpenAPI:
  ```yaml
    Foo:
       type: object
       examples:
         simple-foo-example:
           summary: A simple example of type foo
           externalValue: 'examples/simple-foo-example.json'
       properties:
         createdAt: 
           type: string
           format: date-time
  ```

### PR DESCRIPTION
**Pull request by timotheeguerin**
Fix #2700 Add support for `@example` and `@opExample`

[Playground demo](https://cadlplayground.z22.web.core.windows.net/prs/3572/?c=aW1wb3J0ICJAdHlwZXNwZWMvaHR0cCI7Cgp1c2luZyBIdHRwOwoKQGV4YW1wbGUoI3sKICBpZDogInNvbWUiLAogIGJhcjogInRoaW5nxRBkYXRlOiB1dGNEYXRlVGltZS5mcm9tSVNPKCIyMDIwLTAxLTAxVDAwOsUDWiIpxDV1bml4xDA6xQrENnN0YW1wMzLfPcU9ZW5jb2RlZEFzUmZjNzIzMd9%2B0kFkb2I6IHBsYWluxDbUMsYodGltZW91dDogZHVyYXRpb27KK1BUMU3NJUluU2Vjb25kc9guMS41U9YyRmxvYddlMMc1fSkKbW9kZWwgRm9vIOgBlHN0cmluZzvoAZTKD%2FEBk8QV%2BQFzOwoKICBA5gFVKMgyS25vd25FxBRpbmcucuYBaSn%2FAX1lxUvuAV7Ee%2FEBSM1zInPmAQciLCBpbnQzMsRm%2BgFV2Dtm5AFF1j3vAWA7Cn0K5wFOUGV08wFObmFtZcwQ8ADRfQoKQG9wRecDIwogIOUDJiAgcGFyYW1ldGVyczrIE%2FADPcUnZXTLINQi6ACDIkZsdWZmecsY%2FwK66QK6xAF9xgnHB3JldHVyblR5cGX9AJ3Wed93y3fFbsUFI3sgdGl0bMVOaXJzdCIsIGRlc2NyaXDkAYU6ICJTaG93IGNyZWF05AR6YeQBGCIgfQopCm9wxhdlUGV0KOoBmyzEIjrkAbQpxQY7Cg%3D%3D&e=%40typespec%2Fopenapi3&options=%7B%7D)

### CODE REVIEW

File: packages/openapi3/src/schema-emitter.ts [hunk 1]
```diff
@@ -760,6 +762,17 @@ export class OpenAPI3SchemaEmitter extends TypeEmitter<
     }
   }
 
+  #applySchemaExamples(
```
Comment (timotheeguerin) [inline packages/openapi3/src/schema-emitter.ts:765]: need test for openapi3

File: .chronus/changes/examples-2024-5-14-18-46-26.md [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+          dob: plainDate.fromISO("2020-01-01"),
+        },
+      },
+      returnType: #{
+        id: "some",
+        name: "Fluffy",
+        dob: plainDate.fromISO("2020-01-01"),
+      },
+    },
+    #{ title: "First", description: "Show creating a pet" }
```
Comment (timotheeguerin) [inline .chronus/changes/examples-2024-5-14-18-46-26.md:40]: wondering here if we want to combine the parameters as one so you call this way instead

```
 @opExample(#{
      parameters: #{
        pet: #{
          id: "some",
          name: "Fluffy",
          dob: plainDate.fromISO("2020-01-01"),
        },
      },
      returnType: #{
        id: "some",
        name: "Fluffy",
        dob: plainDate.fromISO("2020-01-01"),
      },

   title: "First", 
   description: "Show creating a pet",
 })
```

Comment (markcowl) [reply .chronus/changes/examples-2024-5-14-18-46-26.md:40]: Although it is wordier, I think putting example metadata in a separate model example is a bit clearer

File: packages/compiler/lib/std/decorators.tsp [hunk 3]
```diff
@@ -500,6 +500,66 @@ extern dec encode(
...
+ * ```tsp
+ * @example(#{name: "Fluffy", age: 2})
+ * model Pet {
+ *  name: string;
+ *  age: int32;
+ * }
+ * ```
+ */
+extern dec example(
+  target: Model | Enum | Scalar | Union | ModelProperty,
```
Comment (markcowl) [inline packages/compiler/lib/std/decorators.tsp:529]: UnionVariant?

File: packages/compiler/src/lib/examples.ts [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+}
+
+function serializeObjectValueAsJson(
+  program: Program,
+  value: ObjectValue,
+  type: Model
+): Record<string, unknown> {
+  const obj: Record<string, unknown> = {};
+  for (const propValue of value.properties.values()) {
+    const definition = type.properties.get(propValue.name);
```
Comment (markcowl) [inline packages/compiler/src/lib/examples.ts:45]: skip property values that do not occur in the type definition?

File: packages/compiler/test/decorators/examples.test.ts [hunk 5]
```diff
@@ -0,0 +1,361 @@
...
+        code: "unassignable",
+      });
+    });
+  });
+
+  describe("scalar", () => {
+    it("valid", async () => {
+      const { program, examples, target } = await getExamplesFor(`
+      @example(test.fromISO("11:32"))
+      @test scalar test extends utcDateTime;
```
Comment (markcowl) [inline packages/compiler/test/decorators/examples.test.ts:95]: should have an example of each valid scalar type

File: packages/compiler/lib/std/decorators.tsp [hunk 6]
```diff
@@ -500,6 +500,66 @@ extern dec encode(
...
+model ExampleOptions {
+  /** The title of the example */
+  title?: string;
+
+  /** Description of the example */
+  description?: string;
+}
+
+/**
+ * Provide an example value for a type.
```
Comment (bterlson) [inline packages/compiler/lib/std/decorators.tsp:513]:
```suggestion
 * Provide an example value for a data type.
```

File: packages/compiler/lib/std/decorators.tsp [hunk 7]
```diff
@@ -500,6 +500,66 @@ extern dec encode(
...
+model OperationExample {
+  /** Example request body. */
+  parameters?: unknown;
+
+  /** Example response body. */
+  returnType?: unknown;
+}
+
+/**
+ * Provide an example value for a type.
```
Comment (bterlson) [inline packages/compiler/lib/std/decorators.tsp:546]:
```suggestion
 * Provide an example value for an operation's parameters and corresponding return type.
```

File: packages/compiler/src/lib/examples.ts [hunk 8]
```diff
@@ -0,0 +1,173 @@
+import { Temporal } from "temporal-polyfill";
+import type { Program } from "../core/program.js";
+import type { Model, ObjectValue, Scalar, ScalarValue, Type, Value } from "../core/types.js";
+import { getEncode, type EncodeData } from "./decorators.js";
+
+/**
+ * Serialize the given TypeSpec value as a JSON object using the given type and its encoding annotations.
+ * The Value MUST be assignable to the given type.
+ */
+export function serializeValueAsJson(
```
Comment (bterlson) [inline packages/compiler/src/lib/examples.ts:10]: Should this be generalized as any serializing to values for any content type perhaps? Little bit concerned about a lot of JSON-specific logic here.

**Review by bterlson — APPROVED** [review]

File: packages/compiler/test/decorators/examples.test.ts (hunk 5 above)
Comment (timotheeguerin) [reply packages/compiler/test/decorators/examples.test.ts:95]: there is test cases below that go over serialization for each scalar type, this was just to test you can apply to a scalar

File: packages/compiler/src/lib/examples.ts (hunk 8 above)
Comment (timotheeguerin) [reply packages/compiler/src/lib/examples.ts:10]: talked offline, this is good for now and we can see to design  a more generic system where you can register different serializer

**Review by markcowl — APPROVED** [review]

**Pull request merged**